src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchWriter.cs

```diff
@@ -462,5 +658,95 @@ private void WritePendingMessageData(bool reportMessageCompleted)
+                    this.RawOutputContext.TextWriter,
+                    this.batchBoundary,
+                    !this.batchStartBoundaryWritten).ConfigureAwait(false);
+                this.batchStartBoundaryWritten = true;
+            }
+            else
+            {
+                await ODataMultipartMixedBatchWriterUtils.WriteStartBoundaryAsync(
+                    this.RawOutputContext.TextWriter,
+                    this.changeSetBoundary,
+                    !this.changesetStartBoundaryWritten).ConfigureAwait(false);
+                this.changesetStartBoundaryWritten = true;
+            }
+        }
+
+        /// <summary>
+        /// Asynchronously write any pending headers for the current operation message (if any).
+        /// </summary>
+        /// <param name="reportMessageCompleted">
+        /// A flag to control whether after writing the pending data we report writing the message to be completed or not.
+        /// </param>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        private async Task WritePendingMessageDataAsync(bool reportMessageCompleted)
+        {
+            if (this.CurrentOperationMessage != null)
+            {
+                Debug.Assert(this.RawOutputContext.TextWriter != null, "Must have a batch writer if pending data exists.");
+
+                if (this.CurrentOperationResponseMessage != null)
+                {
+                    Debug.Assert(this.RawOutputContext.WritingResponse, "If the response message is available we must be writing response.");
+                    int statusCode = this.CurrentOperationResponseMessage.StatusCode;
+                    string statusMessage = HttpUtils.GetStatusMessage(statusCode);
+
+                    await this.RawOutputContext.TextWriter.WriteLineAsync(
+                        string.Concat(ODataConstants.HttpVersionInBatching, " ", statusCode, " ", statusMessage)).ConfigureAwait(false);
+                }
+
+                IEnumerable<KeyValuePair<string, string>> headers = this.CurrentOperationMessage.Headers;
+                if (headers != null)
+                {
+                    foreach (KeyValuePair<string, string> headerPair in headers)
+                    {
+                        await this.RawOutputContext.TextWriter.WriteLineAsync(
+                            string.Concat(headerPair.Key, ": ", headerPair.Value))
+                            .ConfigureAwait(false);
+                    }
+                }
+
+                // Write CRLF after the headers (or request/response line if there are no headers)
+                await this.RawOutputContext.TextWriter.WriteLineAsync()
+                    .ConfigureAwait(false);
+
+                if (reportMessageCompleted)
+                {
+                    this.CurrentOperationMessage.PartHeaderProcessingCompleted();
+                    this.CurrentOperationRequestMessage = null;

```

Comment:
Do we need to check if current operationrequest message is null? Is this method always called once?

Reply:
This logic is inside an `if (this.CurrentOperationMessage != null)` block. Here's a stripped down version of  `CurrentOperationMessage` property definition:
```csharp
private ODataBatchOperationMessage CurrentOperationMessage
{
    get
    {
        if (this.CurrentOperationRequestMessage != null)
        {
            return this.CurrentOperationRequestMessage.OperationMessage;
        }
        else if (this.CurrentOperationResponseMessage != null)
        {
            return this.CurrentOperationResponseMessage.OperationMessage;
        }
        else
        {
            return null;
        }
    }
}
```
The call to `this.CurrentOperationMessage.PartHeaderProcessingCompleted()` should be safe since it'll never be triggered unless either `CurrentOperationRequestMessage` or `CurrentOperationResponseMessage` is non-null.
Assigning null to both `CurrentOperationRequestMessage` and `CurrentOperationResponseMessage` is also safe

Reply:
Thanks @gathogojr